moodle/Sniffs/PHPUnit/ParentSetUpTearDownSniff.php

```diff
@@ -0,0 +1,222 @@
+            while ($mStart = $phpcsFile->findNext(T_FUNCTION, $mStart + 1, $tokens[$cStart]['scope_closer'])) {
+                $method = $phpcsFile->getDeclarationName($mStart);
+                // Only interested in setUp and tearDown methods.
+                if (!in_array($method, $allMethods)) {
+                    continue;
+                }
+
+                // Iterate over all the parent:: calls in the method.
+                $pStart = $mStart;
+                $correctParentCalls = [];
+                while ($pStart = $phpcsFile->findNext(T_PARENT, $pStart + 1, $tokens[$mStart]['scope_closer'])) {
+                    // The next-next token should be the parent method being named.
+                    $methodCall = $phpcsFile->findNext(T_STRING, $pStart + 1, $pStart + 3);
+                    // If we are calling to an incorrect parent method, report it. No fixable.
+                    if (
+                        $methodCall !== false &&
+                        $tokens[$methodCall]['content'] !== $method &&
+                        in_array($tokens[$methodCall]['content'], $allMethods) // Other parent calls may be correct.
+                    ) {
+                        $wrongMethod = $tokens[$methodCall]['content'];
+                        // We are calling to incorrect parent method.
+                        $phpcsFile->addError(
+                            'The %s() method in unit tests cannot call to parent::%s().',
+                            $pStart,
+                            'Incorrect' . ucfirst($method),
+                            [$method, $wrongMethod]
+                        );
+                    }
+
+                    // If we are calling to the correct parent method, annotate it.
+                    if (
+                        $methodCall !== false &&
+                        $tokens[$methodCall]['content'] === $method
+                    ) {
+                        $correctParentCalls[] = $pStart;
+                    }
+                }
+
+                // If there are multiple calls to correct parent, report it. Not fixable.
+                if (count($correctParentCalls) > 1) {
+                    $phpcsFile->addError(
+                        'The %s() method in unit tests must call to parent::%s() only once.',
+                        end($correctParentCalls),
+                        'Multiple' . ucfirst($method),
+                        [$method, $method]
+                    );
+                }
+
+                // If there are no calls to correct parent, report it. Fixable.
+                if (count($correctParentCalls) === 0) {
+                    // Any weird case where the method is empty, we need at very least 1 line. Skip.
+                    $startLine = $tokens[$tokens[$mStart]['scope_opener']]['line'];
+                    $endLine = $tokens[$tokens[$mStart]['scope_closer']]['line'];
+                    if ($startLine === $endLine) {
```

Comment:
Rather than this, perhaps something like:

```php
// Find the next thing that is not an empty token.
$ignore = \PHP_CodeSniffer\Util\Tokens::$emptyTokens;

$nextValidStatement = $phpcsFile->findNext($ignore, $mStart, $tokens[$mStart]['scope_closer'], true);
if ($nextValidStatement === false) {
    continue;
}
```

Reply:
Okidoki, I like it. Will apply soon.